genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/actions/CleanupJobAction.java

```diff
@@ -48,6 +57,50 @@
     protected Events executeStateAction(final ExecutionContext executionContext) {
         log.info("Cleaning up job...");
 
+        // Set if this job got past the state where a job ID is successfully claimed.
+        final String claimedJobId = executionContext.getClaimedJobId();
+
+        // Set if the job was launched, null if that state was never reached,
+        // for example due to CANCEL_JOB_LAUNCH event.
+        final JobStatus finalJobStatus = executionContext.getFinalJobStatus();
+
+        // Reason for the job being killed.
+        final ExecutionContext.KillProvenance killProvenance = executionContext.getJobKillProvenance();
+
+        // Last job status the server was made aware of.
+        final JobStatus lastJobStatus = executionContext.getCurrentJobStatus();
+
+        if (lastJobStatus != null && claimedJobId != null) {
+            // A job was claimed, but final status is not set (due to error/kill/...).
+            // server should be made aware of final state.
+            if (finalJobStatus == null) {
+                // This job was killed
+                if (killProvenance != null) {
+                    try {
+                        // Job launch was aborted before the the job even started.
+                        agentJobService.changeJobStatus(
+                            claimedJobId,
+                            lastJobStatus,
+                            JobStatus.KILLED,
+                            "Terminated by user via " + killProvenance.name()
+                        );
+                    } catch (final ChangeJobStatusException e) {
+                        throw new RuntimeException("Failed to update server status", e);
+                    }
+                    executionContext.setFinalJobStatus(JobStatus.KILLED);
+                } else {
+                    // No job final status (which reflect server-side status), and no sign of abort/kill.
+                    throw new IllegalStateException(
+                        "Reached cleanup state and finalJobState is null. Last job state: " + lastJobStatus
+                    );
+                }
+            } else {
+                log.debug("Job final status already updated server-side: {}", finalJobStatus);
+            }
+        } else {
+            log.debug("Job never claimed an ID, skipping server-side status update");
```

Comment:
This can happen even if `lastJobStatus == null`. The log statement does not represent that

Reply:
It does.
Not having reached 'claimed' status is the important bit, it covers a range of states, including no state ever set.